security/pkg/util/k8sutil.go

```diff
@@ -0,0 +1,155 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	k8sauth "k8s.io/api/authentication/v1"
+)
+
+type specForSaValidationRequest struct {
+	Token string `json:"token"`
+}
+
+type saValidationRequest struct {
+	APIVersion string                     `json:"apiVersion"`
+	Kind       string                     `json:"kind"`
+	Spec       specForSaValidationRequest `json:"spec"`
+}
+
+// reviewServiceAccountAtK8sAPIServer reviews the CSR credential (k8s service account) at k8s API server.
+// k8sAPIServerURL: the URL of k8s API Server
+// k8sAPIServerCaCert: the CA certificate of k8s API Server
+// reviewerToken: the service account of the k8s token reviewer
+// jwt: the JWT of the k8s service account
+func reviewServiceAccountAtK8sAPIServer(k8sAPIServerURL string, k8sAPIServerCaCert []byte,
+	reviewerToken string, jwt []byte) (*http.Response, error) {
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(k8sAPIServerCaCert)
+	saReq := saValidationRequest{
+		APIVersion: "authentication.k8s.io/v1",
+		Kind:       "TokenReview",
+		Spec:       specForSaValidationRequest{Token: string(jwt[:])},
+	}
+	saReqJSON, err := json.Marshal(saReq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal the service account review request: %v", err)
+	}
+	req, err := http.NewRequest("POST", k8sAPIServerURL, bytes.NewBuffer(saReqJSON))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a HTTP request: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "Bearer "+reviewerToken)
+	// Set the TLS certificate
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: caCertPool,
+			},
+		},
+	}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send the HTTP request: %v", err)
+	}
+	return resp, nil
+}
+
+// ValidateK8sJwt validates a k8s JWT at API server.
+// Return <namespace>:<serviceaccountname> in the JWT when the validation passes.
+// Otherwise, return the error.
+// k8sAPIServerURL: the URL of k8s API Server
+// k8sAPIServerCaCert: the CA certificate of k8s API Server
+// reviewerToken: the service account of the k8s token reviewer
+// jwt: the JWT to validate
+func ValidateK8sJwt(k8sAPIServerURL string, k8sAPIServerCaCert []byte, reviewerToken string,
```

Comment:
Please try encapsulate this function with a structure. This will support polymorphism and facilitate mocking at the caller. See https://github.com/istio/istio/pull/10657/files#diff-3b1228513d1272dcb0529094dd8fbd7b

Reply:
Done.